Able to load a uds file to pointCloud

kressnick25 opened the pull request:
Implemented loading `uds` file to `pointCloud` with file selector

Co-authored by @jmkraut 
## Other notable changes
- Handled errors better using exceptions by extending a parent UdSDKClass which all udSDK classes extend and therefore share the same error handling.

- Pointers and allocations now handled internally by each UdSDK class so as to avoid doing memory management outside lib. Just have to call cleanup() to free allocated memory. just have to be careful not to allow object to be de-referenced before calling cleanup() as this will leak memory. 

- Fixed the launch config so don't have to launch twice
## Screenshots
### Home page
![Screenshot_20200827-154227](https://user-images.githubusercontent.com/35619338/91389209-34581d00-e87c-11ea-8e62-58aea91e8902.png)
### Select file screen
![Screenshot_20200827-154236](https://user-images.githubusercontent.com/35619338/91389215-3621e080-e87c-11ea-87cb-c511a036a4cf.png)
### Error message for bad file format
Can be changed to more descriptive, this is just to show it works and doesn't crash the application on failure
![Screenshot_20200827-154250](https://user-images.githubusercontent.com/35619338/91389219-37530d80-e87c-11ea-9ed9-fefc1db6f24d.png)

kressnick25 committed:
Added register button + pf logo to login screen

kressnick25 committed:
Implemented option to disable cert security

kressnick25 committed:
Fixed bug caused by ignoreCert not properly being set

kressnick25 committed:
Prevented screen from rotating

kressnick25 committed:
Login redirects to new Home screen

kressnick25 committed:
updated comments

kressnick25 committed:
Fixed screens being wrapped with MaterialApp

Only the main app wrapper should be wrapped with MaterialApp as this
allows context and theming to be defined once and then inherited by all
child widgets.

kressnick25 committed:
added helper for getting value of udError code

kressnick25 committed:
Implemented udContext_Disconnect binding

kressnick25 committed:
added udAttributes bindings

Commit:
Integrated a "select file" option and allow an absolute path to be selected from the device.

kressnick25 committed:
UdContext stores context pointer as internal state

Doing this allows us to handle memory allocations in the library

kressnick25 committed:
Get address from udContext

kressnick25 committed:
handle errors from UdSDK classes

Throws exception if error has occurs with proper error message

kressnick25 committed:
updated error handling in udConfig

kressnick25 committed:
updated error handling in udAttributes

kressnick25 committed:
fixed missing ArrayHelper for udAttributes

kressnick25 committed:
removed unused imports

kressnick25 committed:
updated error handling in login screen

kressnick25 committed:
fixed type definitions

kressnick25 committed:
udPointCloud bindings

kressnick25 committed:
launch config

kressnick25 committed:
Merge branch 'file_selector' into udContext-bindings

kressnick25 committed:
Error display widget

kressnick25 committed:
arrayHelper to copy C array to Dart List

kressnick25 committed:
changed header from pointer to internal value

kressnick25 committed:
able to call udPointCloud_Load on uds file

samsinsane commented on vaultSDK/lib/udAttributes.dart:
```diff
@@ -0,0 +1,730 @@
+  /// @param pOffset This pointer will be written to with the value of the offset if it is found
+  /// @return A udError value based on the result of writing the offset to pOffset
+  void getOffsetOfNamedAttribute(String name, int offset) {
+    final Pointer<Uint32> pOffset = allocate();
```
You don't call `free` for this.

kressnick25 replied:
Are we doing this completely wrong then? I just looked at what the function does and I'm thinking we should either:
- return an `int` from the wrapper function `getOffsetNamedAttribute`
- leave the return type as `void` and have `Pointer<Uint32> pOffset` as a property of the class

samsinsane replied:
My apologies, I didn't look very closely at the function itself, just that it differed to the one above. Yes, this should return an `int` (the offset), so it should look something like:
```dart
int getOffsetOfNamedAttribute(String name) {
  final Pointer<Uint32> pOffset = allocate();
  final pName = Utf8.toUtf8(name);
  handleUdError(_udAttributeSet_GetOffsetOfNamedAttribute(
      _attributeSet, pName, pOffset));
  final offset = pOffset.value;
  free(pName);
  free(pOffset);
  return offset;
}
```

samsinsane replied:
Also, to clarify, the function above this one should be adjusted similarly.

kressnick25 replied:
Updated now 262d68d

kressnick25 committed:
changed getOffsetOf...Atribute to return the offset int

Also verfied the behavior of try/catch/finally. Finally will always be
executed regardless of returns/exceptions thrown in preceding try/catch
